p5/core/PShape.py

```diff
@@ -0,0 +1,157 @@
+#py#
+# Part of p5: A Python package based on Processing
+# Copyright (C) 2017 Abhik Pal
+# Copyright (C) 2018 Aliabbas Merchant
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful, but
+# WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
+# General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program. If not, see <http://www.gnu.org/licenses/>.
+#
+
+from . import primitives
+from ..pmath import Point, Vector
+import warnings
+
+MIN_POINT_ACCURACY = 20
+MAX_POINT_ACCURACY = 200
+POINT_ACCURACY_FACTOR = 10
+
+OPEN, CLOSE = 1, 2
+GROUP = 0  # createShape()
+POINT = 2  # primitive
+POINTS = 3  # vertices
+LINE = 4  # primitive
+LINES = 5  # beginShape(), createShape()
+LINE_STRIP = 50  # beginShape()
+LINE_LOOP = 51
+TRIANGLE = 8  # primitive
+TRIANGLES = 9  # vertices
+TRIANGLE_STRIP = 10  # vertices
+TRIANGLE_FAN = 11  # vertices
+QUAD = 16  # primitive
+QUADS = 17  # vertices
+QUAD_STRIP = 18  # vertices
+POLYGON = 20  # in the end, probably cannot
+PATH = 21  # separate these two
+RECT = 30  # primitive
+ELLIPSE = 31  # primitive
+ARC = 32  # primitive
+SPHERE = 40  # primitive
+BOX = 41  # primitive
+
+
+class PShape():
+    def __init__(self, vertices, kind='POLY', p=None):
+        self._stroke = True
+        self._stroke_color = (1,1,1)
+        self._filled = True
+        self._fill = (0, 0, 0)
+        self._visible = True
+        self._openShape = False
+        self._kind = type
+        self._p = p
+        self._vertices = []
+        self.close = False
+
+    def beginShape(self, kind = None):
```

Comment:
We're trying to use context managers in p5. So instead of 
```
foo.beginShape()
# code
foo.endShape()
```
we would like to have
```
with foo.begin_shape():
    # code
````
See the code for ```push_matrix()``` in p5/transforms.py for the implementation details. And the contextlib docs: https://docs.python.org/3/library/contextlib.html

Reply:
Yes.
That would be more pythonic.
My main aim was initially just to make it function just like the Java code.

Reply:
>  My main aim was initially just to make it function just like the Java code.

We don't have to. We're just keeping the higher-level api similar, the internals can definitely change! And of course, we always try to use some of the Python features whenever possible :)

Reply:
Ok.  I'll keep that in mind.  
This is the same reason why I used `beginShape()` instead of `begin_shape()`

Reply:
:+1: